math/src/elliptic_curve/short_weierstrass/curves/bls12_381/twist.rs

```diff
@@ -46,7 +47,57 @@ impl IsShortWeierstrass for BLS12381TwistCurve {
     }
 }
 
+impl BLS12381TwistCurve {
+    pub fn is_on_curve(point: &<Self as IsEllipticCurve>::PointRepresentation) -> bool {
+        if point.z() == &FieldElement::zero() {
+            true
+        } else {
+            let point = point.to_affine();
+            Self::defining_equation(&point.x(), &point.y()) == FieldElement::zero()
+        }
+    }
+}
+
 impl ShortWeierstrassProjectivePoint<BLS12381TwistCurve> {
+    /// Returns the "Untwist-Frobenius-Twist" endomorphism
+    pub fn psi(&self) -> Self {
+        // Coefficients borrowed from https://github.com/Consensys/gnark-crypto/blob/55489ac07ca1a88bf0b830a29625fcb0d8879a48/ecc/bls12-381/bls12-381.go#L132C1-L135C138
+        let psi_coeff_x = FieldElement::<<BLS12381TwistCurve as IsEllipticCurve>::BaseField>::new([
+            FieldElement::zero(),
+            FieldElement::new(U384::from_hex_unchecked("1A0111EA397FE699EC02408663D4DE85AA0D857D89759AD4897D29650FB85F9B409427EB4F49FFFD8BFD00000000AAAD"))
+        ]);
+
+        let psi_coeff_y = FieldElement::<<BLS12381TwistCurve as IsEllipticCurve>::BaseField>::new([
+            FieldElement::new(U384::from_hex_unchecked("135203E60180A68EE2E9C448D77A2CD91C3DEDD930B1CF60EF396489F61EB45E304466CF3E67FA0AF1EE7B04121BDEA2")),
+            FieldElement::new(U384::from_hex_unchecked("6AF0E0437FF400B6831E36D6BD17FFE48395DABC2D3435E77F76E17009241C5EE67992F72EC05F4C81084FBEDE3CC09"))
+        ]);
+
+        let frob_map_x = Degree2ExtensionField::frobenius_map(self.x().value());
+        let frob_map_x_with_psi_coeff_x =
+            Degree2ExtensionField::mul(&frob_map_x, psi_coeff_x.value());
+
+        let frob_map_y = Degree2ExtensionField::frobenius_map(self.y().value());
+        let frob_map_y_with_psi_coeff_y =
+            Degree2ExtensionField::mul(&frob_map_y, psi_coeff_y.value());
+
+        let frob_map_z = Degree2ExtensionField::frobenius_map(self.z().value());
+
+        <BLS12381TwistCurve as IsEllipticCurve>::PointRepresentation::new([
+            FieldElement::new(frob_map_x_with_psi_coeff_x),
+            FieldElement::new(frob_map_y_with_psi_coeff_y),
+            FieldElement::new(frob_map_z),
+        ])
+    }
+
+    /// Returns true if the point is in the prime subgroup `G_2` of order `r`.
+    /// Makes use of endomorphism for efficient point multiplication.
+    pub fn is_in_subgroup(&self) -> bool {
+        let seed_times_p = self.operate_with_self(SEED);
+        let psi_plus_seed_times_p = self.psi().operate_with(&seed_times_p);
+
+        BLS12381TwistCurve::is_on_curve(&psi_plus_seed_times_p)
```

Comment:
This is not checking that $\psi(P) - uP$ is the neutral element. You probably mean something different here.

```suggestion
        psi_plus_seed_times_p.is_neutral_element()
```